File: backend/app/schemas/exercise.py
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.word import WordOut


class ExerciseSettingsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    exercise_key: str
    word_count: int
    # Only meaningful for exercises that use them (currently "build_word");
    # null for every other exercise_key.
    wrong_letter_count: int | None = None
    min_word_length: int | None = None
    case_sensitive: bool | None = None
    # How much a right/wrong answer to this exercise moves a word's
    # WordProgress.score (see app/models/word_progress.py). Null means "use
    # this exercise_key's own code-level default" (see
    # app/exercises/common.py's DEFAULT_POINTS), same fallback pattern as
    # the fields above.
    correct_points: int | None = None
    incorrect_points: int | None = None
    # Whether this exercise_key can be picked for a new Lesson at all. Null
    # means enabled (same "unset = working default" convention as every
    # other field here) -- see app/exercises/__init__.py.
    enabled: bool | None = None
    # Only meaningful for "listen_word" (how many word choices one round
    # shows); null for every other exercise_key.
    option_count: int | None = None
    # Only meaningful for "speaking_word" (0-100 minimum text-similarity to
    # accept a spoken answer); null for every other exercise_key.
    speech_match_threshold: int | None = None


class ExerciseSettingsIn(BaseModel):
    word_count: int = Field(ge=1, le=100)
    wrong_letter_count: int | None = Field(default=None, ge=0, le=20)
    min_word_length: int | None = Field(default=None, ge=1, le=50)
    case_sensitive: bool | None = None
    correct_points: int | None = Field(default=None, ge=0, le=100)
    incorrect_points: int | None = Field(default=None, ge=0, le=100)
    enabled: bool | None = None
    option_count: int | None = Field(default=None, ge=2, le=10)
    speech_match_threshold: int | None = Field(default=None, ge=0, le=100)


class TrueOrFalseItemOut(BaseModel):
    word_id: int
    original: str
    transcription: str | None
    image_url: str | None
    word_audio_url: str | None
    # Always the text actually shown on the card -- either the word's own
    # real translation, or one borrowed from a different learned word.
    shown_translation: str
    shown_translation_audio_url: str | None
    # True iff `shown_translation` is this word's own real translation.
    is_correct: bool


class TrueOrFalseRoundOut(BaseModel):
    dictionary_id: int
    # How many learned (and usable -- i.e. with at least one translation)
    # words existed to draw from. len(items) is min(admin word_count,
    # available_count); the client can use this to tell "fewer words than
    # configured" apart from "the configured count in full".
    available_count: int
    items: list[TrueOrFalseItemOut]


class ExerciseWordsOut(BaseModel):
    """The generic shape for any exercise that just needs N already-learned
    Word rows and does its own thing with them client-side (e.g.
    "Сопоставление", which only needs the words themselves to shuffle into
    two columns) -- no per-item decision like True/False's real-or-fake
    translation choice."""

    dictionary_id: int
    exercise_key: str
    available_count: int
    words: list[WordOut]


class BuildWordItemOut(BaseModel):
    word_id: int
    translation: str
    # The word to build, exactly as stored (real casing/duplicates
    # preserved) -- the client compares the assembled letters against this
    # position by position; whether that comparison is case-sensitive is
    # `BuildWordRoundOut.case_sensitive`, not decided per item.
    correct_word: str
    # Every letter the player can tap: the word's own letters (with their
    # real multiplicity) plus the configured number of wrong distractor
    # letters, already shuffled together. Never fewer distractors than
    # configured only because the alphabet couldn't supply enough distinct
    # ones -- see get_build_word_round.
    letters: list[str]
    transcription: str | None
    image_url: str | None
    word_audio_url: str | None
    translation_audio_url: str | None


class BuildWordRoundOut(BaseModel):
    dictionary_id: int
    available_count: int
    case_sensitive: bool
    items: list[BuildWordItemOut]

File: mobile/lib/models/exercise.dart
/// One card of a "Правда или ложь" round. The backend already decided
/// whether [shownTranslation] is the word's real translation or one
/// borrowed from a different learned word -- [isCorrect] is that decision,
/// never recomputed here. The UI only renders this and compares the
/// user's tap to it.
class TrueOrFalseItem {
  final int wordId;
  final String original;
  final String? transcription;
  final String? imageUrl;
  final String? wordAudioUrl;
  final String shownTranslation;
  final String? shownTranslationAudioUrl;
  final bool isCorrect;

  TrueOrFalseItem({
    required this.wordId,
    required this.original,
    this.transcription,
    this.imageUrl,
    this.wordAudioUrl,
    required this.shownTranslation,
    this.shownTranslationAudioUrl,
    required this.isCorrect,
  });

  factory TrueOrFalseItem.fromJson(Map<String, dynamic> json) {
    return TrueOrFalseItem(
      wordId: json['word_id'] as int,
      original: json['original'] as String,
      transcription: json['transcription'] as String?,
      imageUrl: json['image_url'] as String?,
      wordAudioUrl: json['word_audio_url'] as String?,
      shownTranslation: json['shown_translation'] as String,
      shownTranslationAudioUrl: json['shown_translation_audio_url'] as String?,
      isCorrect: json['is_correct'] as bool,
    );
  }
}

class TrueOrFalseRound {
  final int dictionaryId;
  final int availableCount;
  final List<TrueOrFalseItem> items;

  TrueOrFalseRound({required this.dictionaryId, required this.availableCount, required this.items});

  factory TrueOrFalseRound.fromJson(Map<String, dynamic> json) {
    return TrueOrFalseRound(
      dictionaryId: json['dictionary_id'] as int,
      availableCount: json['available_count'] as int,
      items: (json['items'] as List<dynamic>)
          .map((e) => TrueOrFalseItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// One "Собери слово" task. [letters] is already the full shuffled set
/// (real letters + distractors) the backend decided -- nothing about
/// which/how many wrong letters, or the shuffle, happens on-device.
/// [correctWord] is used only for local, immediate comparison as the
/// player builds their answer; the backend never re-checks it.
class BuildWordItem {
  final int wordId;
  final String translation;
  final String correctWord;
  final List<String> letters;
  final String? transcription;
  final String? imageUrl;
  final String? wordAudioUrl;
  final String? translationAudioUrl;

  BuildWordItem({
    required this.wordId,
    required this.translation,
    required this.correctWord,
    required this.letters,
    this.transcription,
    this.imageUrl,
    this.wordAudioUrl,
    this.translationAudioUrl,
  });

  factory BuildWordItem.fromJson(Map<String, dynamic> json) {
    return BuildWordItem(
      wordId: json['word_id'] as int,
      translation: json['translation'] as String,
      correctWord: json['correct_word'] as String,
      letters: (json['letters'] as List<dynamic>).map((e) => e as String).toList(),
      transcription: json['transcription'] as String?,
      imageUrl: json['image_url'] as String?,
      wordAudioUrl: json['word_audio_url'] as String?,
      translationAudioUrl: json['translation_audio_url'] as String?,
    );
  }
}

class BuildWordRound {
  final int dictionaryId;
  final int availableCount;
  final bool caseSensitive;
  final List<BuildWordItem> items;

  BuildWordRound({
    required this.dictionaryId,
    required this.availableCount,
    required this.caseSensitive,
    required this.items,
  });

  factory BuildWordRound.fromJson(Map<String, dynamic> json) {
    return BuildWordRound(
      dictionaryId: json['dictionary_id'] as int,
      availableCount: json['available_count'] as int,
      caseSensitive: json['case_sensitive'] as bool,
      items: (json['items'] as List<dynamic>)
          .map((e) => BuildWordItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}
